distribution/tools/plugin-cli/src/main/java/org/elasticsearch/plugins/InstallPluginCommand.java

```diff
@@ -253,6 +242,60 @@ private Path download(Terminal terminal, String pluginId, Path tmpDir) throws Ex
         return downloadZip(terminal, pluginId, tmpDir);
     }
 
+    // pkg private so tests can override
+    String getStagingHash() {
+        return System.getProperty(PROPERTY_STAGING_ID);
+    }
+
+    /** Returns the url for an official elasticsearch plugin. */
+    private String getElasticUrl(Terminal terminal, String stagingHash, Version version,
+                                        String pluginId, String platform) throws IOException {
+        final String baseUrl;
+        if (stagingHash != null) {
+            baseUrl = String.format(Locale.ROOT,
+                "https://staging.elastic.co/%1$s-%2$s/downloads/elasticsearch-plugins/%3$s", version, stagingHash, pluginId);
+        } else {
+            baseUrl = String.format(Locale.ROOT,
+                "https://artifacts.elastic.co/downloads/elasticsearch-plugins/%1$s", pluginId);
+        }
+        final String platformUrl = String.format(Locale.ROOT, "%1$s/%2$s-%3$s-%4$s.zip", baseUrl, pluginId, platform, version);
+        if (urlExists(terminal, platformUrl)) {
+            return platformUrl;
+        }
+        return String.format(Locale.ROOT, "%1$s/%2$s-%3$s.zip", baseUrl, pluginId, version);
+    }
+
+    /** Returns the url for an elasticsearch plugin in maven. */
+    private String getMavenUrl(Terminal terminal, String[] coordinates, String platform) throws IOException {
+        final String groupId = coordinates[0].replace(".", "/");
+        final String artifactId = coordinates[1];
+        final String version = coordinates[2];
+        final String baseUrl = String.format(Locale.ROOT, "https://repo1.maven.org/maven2/%1$s/%2$s/%3$s", groupId, artifactId, version);
+        final String platformUrl = String.format(Locale.ROOT, "%1$s/%2$s-%3$s-%4$s.zip", baseUrl, artifactId, platform, version);
+        if (urlExists(terminal, platformUrl)) {
+            return platformUrl;
+        }
+        return String.format(Locale.ROOT, "%1$s/%2$s-%3$s.zip", baseUrl, artifactId, version);
+    }
+
+    /**
+     * Returns {@code true} if the given url exists, and {@code false} otherwise.
+     *
+     * The given url must be {@code https} and existing means a {@code HEAD} request returns 200.
+     */
+    // pkg private for tests to manipulate
+    @SuppressForbidden(reason = "Make HEAD request using URLConnection.connect()")
+    boolean urlExists(Terminal terminal, String urlString) throws IOException {
+        terminal.println(VERBOSE, "Checking if url exists: " + urlString);
+        assert urlString.startsWith("https://") : "Only http urls can be checked";
+        URL url = new URL(urlString);
+        HttpURLConnection urlConnection = (HttpURLConnection)url.openConnection();
+        urlConnection.addRequestProperty("User-Agent", "elasticsearch-plugin-installer");
+        urlConnection.setRequestMethod("HEAD");
+        urlConnection.connect();
+        return urlConnection.getResponseCode() == 200;
```

Comment:
Ah, okay. Then I'm good with it as is; we can consider redirects separately (if ever).